### PR TITLE
Improve iterating over many field producers during downsample operation

### PR DESCRIPTION
**Pull request by martijnvg**
The downsample operation is a heavy operation. It scans all documents and rolls up all the metric and label fields.
This is done in `TimeSeriesBucketCollector#collect(...)` method. This is hot code, for each document it iterates over all label and metric fields. An immutable list is used for looping, and this loop showed up as hot code. This is especially true when downsampling an already downsampled index (then there are 5 times more metric fields). See flame graph below.

This change uses arrays instead of jdk immutable list.

Relates to #97141

image

### CODE REVIEW

**Comment by martijnvg** [comment]
Running the downsample challenge in the tsdb track showed a good improvement with this change:

```
| 100th percentile service time |              downsample-1m | 738085           | 633226           | -104859       |     ms |  -14.21% |
|                    error rate |              downsample-1m |      0           |      0           |       0       |      % |    0.00% |
| 100th percentile service time |              downsample-1h | 495576           | 446490           |  -49085.5     |     ms |   -9.90% |
|                    error rate |              downsample-1h |      0           |      0           |       0       |      % |    0.00% |
| 100th percentile service time |              downsample-1d | 493133           | 454931           |  -38201.2     |     ms |   -7.75% |
|                    error rate |              downsample-1d |      0           |      0           |       0       |      % |    0.00% |
```

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java [hunk 1]
```diff
@@ -250,9 +249,9 @@ private class TimeSeriesBucketCollector extends BucketCollector {
 
         TimeSeriesBucketCollector(BulkProcessor2 bulkProcessor) {
             this.bulkProcessor = bulkProcessor;
-            List<AbstractDownsampleFieldProducer> rollupFieldProducers = fieldValueFetchers.stream()
+            AbstractDownsampleFieldProducer[] rollupFieldProducers = fieldValueFetchers.stream()
```
Comment (salvatore-campagna) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:252]: I wonder if here...not using stream might give even better results.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java [hunk 2]
```diff
@@ -383,18 +385,18 @@ private class RollupBucketBuilder {
...
+        RollupBucketBuilder(AbstractDownsampleFieldProducer[] rollupFieldProducers) {
             this.rollupFieldProducers = rollupFieldProducers;
             /*
              * The rollup field producers for aggregate_metric_double all share the same name (this is
              * the name they will be serialized in the target index). We group all field producers by
              * name. If grouping yields multiple rollup field producers, we delegate serialization to
              * the AggregateMetricFieldSerializer class.
              */
-            groupedProducers = rollupFieldProducers.stream()
+            groupedProducers = Arrays.stream(rollupFieldProducers)
```
Comment (salvatore-campagna) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:399]: Same here...maybe a simple for loop has better performance.

Comment (martijnvg) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java:399]: I think we can keep the stream usage here and in `TimeSeriesBucketCollector` constructor here. 
This is called once per shard level downsample execution, so it isn't hot code, like in the places I made changes in this PR.

**Review by salvatore-campagna — APPROVED** [review]

**Pull request merged**